servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AbstractH2DuplexHandler.java

```diff
@@ -65,18 +73,51 @@ public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
         }
     }
 
+    final void writeMetaData(ChannelHandlerContext ctx, HttpMetaData metaData, Http2Headers h2Headers,
+                             ChannelPromise promise) {
+        endStream = !mayHaveTrailers(metaData) && isPayloadEmpty(metaData);
+        if (endStream) {
+            closeHandler.protocolPayloadEndOutbound(ctx, promise);
```

Comment:
@Scottmitch `HttpTransportObserverTest.serverHandlerError` helped to discover a race between netty closing the `StreamChannel` (emits `channelInactive`, `channelUnregistered`, and removes the handler) and receiving trailers from the offloaded thread.

So, I made this change: in case of `endStream`, I trigger this callback here and do nothing in `writeTrailers`. IIUC the control flow of `WriteStreamSubscriber`, it will ignore any following writes (see `doWrite`) if the source is already terminated. 
Do you see any issues with existing approach?

Reply:
Discussed with @Scottmitch offline. `WriteStreamSubscriber` will correctly account for the current state and no further writes or terminal signals will be propagated. All good.